ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,354 @@
+    """
+        # TODO:
+    """
+    
+    config = dict(
+        type='countbased',
+        counter_type='PixelCNN',
+        intrinsic_reward_type='add',
+        # learning_rate=1e-3,
+        batch_size=16,
+        bonus_coeffient=0.1,
+        img_height=42,
+        img_width=42,
+        in_channels=1,
+        out_channels=1,
+        n_gated=2,
+        gated_channels=16,
+        head_channels=64,
+        q_level=256,
+    )
+
+    def __init__(
+        self,
+        cfg: dict,
+        device,
+        tb_logger: 'SummaryWriter'
+    ) -> None:  # noqa
+        """
+            # TODO:
+        """
+        super(BaseRewardModel, self).__init__()
+        self.cfg: dict = cfg
+        self._beta = cfg.bonus_coeffient
+        self._counter_type = cfg.counter_type
+        self.device = device
+        self.tb_logger = tb_logger
+        assert self._counter_type in ['SimHash', 'AutoEncoder', 'PixelCNN']
+        # TODO: copy
+        if self._counter_type == 'PixelCNN':
+            print(cfg)
+            self._counter = GatedPixelCNN(
+                in_channels=cfg.in_channels,
+                out_channels=cfg.out_channels,
+                n_gated=cfg.n_gated,
+                gated_channels=cfg.gated_channels,
+                head_channels=cfg.head_channels,
+                q_level=cfg.q_level,
+            )
+            self._counter.to(self.device)
+            self.intrinsic_reward_type = cfg.intrinsic_reward_type
+            self.obs_shape = (cfg.img_height, cfg.img_width)
+            self.index_range = np.arange(cfg.img_height * cfg.img_width)
+            assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+            self.train_data = []
+            self.opt = optim.RMSprop(
+                self._counter.parameters(),
+                momentum=0.9,
+                weight_decay=0.95,
+                eps=1e-4,
+            )
+
+    def _train(self):
+        train_data: list = self.train_data
+        train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+        loss = nn.CrossEntropyLoss()(flattened_logits, target_pixel_loss.long())
+        loss = loss.mean()
+        self.opt.zero_grad()
+        loss.backward()
+        self.opt.step()
+
+    def train(self) -> None:
+        self._train()
+        # self.clear_data()
+
+    def estimate(self, data, t) -> None:
+        """
+            # TODO:
+        """
+        if self._counter_type == 'PixelCNN':
+            obs = self._collect_states(data)
+            obs = torch.stack(obs).to(self.device)
+
+            prob = (self._probs(obs) + 1e-8).sum().item()
+
+            with torch.no_grad():
+                recoding_prob = (self._probs(obs) + 1e-8).sum().item()
+
+            pred_gain = np.log(recoding_prob) - np.log(prob)
+
+            intrinsic_reward = [pow(
+                exp(self.cfg.bonus_coeffient * pow(t+1, -0.5) * pred_gain) - 1, 0.5
```

Comment:
why t+1 not t, would you please give some annotations here?

Reply:
That's easy to see for case t = 0, where the pred_gain has no use if using t not t+1. And I'll add some annotations, thanks.

Reply:
Ok, got it. Thanks.